pkg/operator/storageclasshook_test.go

```diff
@@ -0,0 +1,171 @@
+	}
+}
+
+func withParameters(sc *storagev1.StorageClass, keysAndValues ...string) *storagev1.StorageClass {
+	for i := 0; i < len(keysAndValues); i += 2 {
+		sc.Parameters[keysAndValues[i]] = keysAndValues[i+1]
+	}
+	return sc
+}
+
+func TestStorageClassHook(t *testing.T) {
+	tests := []struct {
+		name        string
+		driver      *opv1.ClusterCSIDriver
+		expectedSC  *storagev1.StorageClass
+		expectError bool
+	}{
+		{
+			name: "no driver config",
+			driver: &opv1.ClusterCSIDriver{
+				Spec: opv1.ClusterCSIDriverSpec{},
+			},
+			expectedSC: sc(),
+		},
+		{
+			name: "driver config with irrelevant type",
+			driver: &opv1.ClusterCSIDriver{
+				Spec: opv1.ClusterCSIDriverSpec{
+					DriverConfig: opv1.CSIDriverConfigSpec{
+						DriverType: opv1.AWSDriverType,
+					},
+				},
+			},
+			expectedSC: sc(),
+		},
+		{
+			name: "driver config with no spec",
+			driver: &opv1.ClusterCSIDriver{
+				Spec: opv1.ClusterCSIDriverSpec{
+					DriverConfig: opv1.CSIDriverConfigSpec{
+						DriverType: opv1.GCPDriverType,
+					},
+				},
+			},
+			expectedSC: sc(),
+		},
+		{
+			name: "driver config with nil KMSKey",
+			driver: &opv1.ClusterCSIDriver{
+				Spec: opv1.ClusterCSIDriverSpec{
+					DriverConfig: opv1.CSIDriverConfigSpec{
+						DriverType: opv1.GCPDriverType,
+						GCP: &opv1.GCPCSIDriverConfigSpec{
+							KMSKey: nil,
+						},
+					},
+				},
+			},
+			expectedSC: sc(),
+		},
+		{
+			name: "with disk-encryption-kms-key in SC",
+			driver: &opv1.ClusterCSIDriver{
+				Spec: opv1.ClusterCSIDriverSpec{
+					DriverConfig: opv1.CSIDriverConfigSpec{
+						DriverType: opv1.GCPDriverType,
+						GCP: &opv1.GCPCSIDriverConfigSpec{
+							KMSKey: &opv1.GCPKMSKeyReference{
+								ProjectID: testProject,
+								Location:  testLocation,
+								KeyRing:   testKeyRing,
+								Name:      testKMSName,
+							},
+						},
+					},
+				},
+			},
+			expectedSC: withParameters(sc(), diskEncryptionKMSKey, getExpectedSCParam(testProject, testLocation, testKeyRing, testKMSName)),
+		},
+		{
+			name: "with disk-encryption-kms-key in SC, default location",
+			driver: &opv1.ClusterCSIDriver{
+				Spec: opv1.ClusterCSIDriverSpec{
+					DriverConfig: opv1.CSIDriverConfigSpec{
+						DriverType: opv1.GCPDriverType,
+						GCP: &opv1.GCPCSIDriverConfigSpec{
+							KMSKey: &opv1.GCPKMSKeyReference{
+								ProjectID: testProject,
+								KeyRing:   testKeyRing,
+								Name:      testKMSName,
+							},
+						},
+					},
+				},
+			},
+			expectedSC: withParameters(sc(), diskEncryptionKMSKey, getExpectedSCParam(testProject, "global", testKeyRing, testKMSName)),
+		},
+	}
+
+	for _, test := range tests {
+		test := test
+		t.Run(test.name, func(t *testing.T) {
+			ccdLister := &fakeCCDLister{test.driver}
+			hook := getKMSKeyHook(ccdLister)
+			testSC := sc()
+
+			err := hook(nil, testSC)
+
+			if err != nil && !test.expectError {
+				t.Errorf("got unexpected error: %s", err)
+			}
+			if err == nil && test.expectError {
```

Comment:
It is hard to understand (from the point-view of someone who reads the code) why we need this "if". After more thinking, it's getting clear that `expectError` is always `false` in these tests, and all we need is a single "if":
```
			if err != nil {
				t.Errorf("got unexpected error: %s", err)
			}
```

The best way to fix it is to add one more test to `tests` array where StorageClass `sc` is created with wrong name, so `Get()` would return an error, and the test would catch it because this test would have `expectError` equal to `true`.

Alternatively, we could remove `expectError` altogether along with that redundant "if". And, as a third option, we could add a comment like "TODO: Add test(s) with expectError = true". This way the reader would get a hint why we have such a redundancy now.

Reply:
Indeed this condition is not used now, I'd go with adding a failure case as suggested to avoid having unused code. Simply adding a parameter to `sc()` func for provisioner name would work. Then add one more test with incorrect provisioner name like: `expectedSC: sc("invalid-provisioner-name")` and set `expectedError: true` for that test.

Reply:
Good feedback, thanks, I added an invalid provisioner test.